R/fwrite.R

```diff
@@ -38,7 +38,8 @@ fwrite = function(x, file="", append=FALSE, quote="auto",
   # validate arguments
   if (is.matrix(x)) { # coerce to data.table if input object is matrix
     messagef("x being coerced from class: matrix to data.table")
-    x = as.data.table(x)
+    # keep row.names for matrix input #5315
+    x = if (row.names && !is.null(rownames(x))) { row.names = FALSE; as.data.table(x, keep.rownames="") } else as.data.table(x)
```

Comment:
Not a fan of inlining assignment to `x` given that one branch has a "side effect" of setting `row.names=FALSE`.

```suggestion
    if (row.names && !is.null(rownames(x))) {
      row.names = FALSE
      x = as.data.table(x, keep.rownames="")
    } else {
      x = as.data.table(x)
    }
```

PS we haven't done any input validation on `row.names` by this point, so maybe this should be `if (isTRUE(row.names) && !is.null(rownames(x))) { ... }`